File: lua/puzzle5-2.lua
require "utils"

local seeds = io.read()
assert(seeds)
seeds = string.split(seeds, ": ")[2]
assert(seeds)
seeds = string.split(string.strip(seeds))
for i in ipairs(seeds) do seeds[i] = seeds[i] + 0 end
local seedrngs = {}
for i = 1, #seeds, 2 do
	local start, len = seeds[i], seeds[i + 1]
	table.insert(seedrngs, {x1=start, x2=start + len - 1})
end
print('seedrngs', dump(seedrngs))
io.read()

local maps = {}
for line in io.lines() do
	local map = {}
	map.name = string.split(line)[1]
	assert(map.name)
	map.ranges = {}
	for rngline in io.lines() do
		if rngline == "" then break end
		local range = string.split(rngline)
		assert(#range == 3)
		local dst, src, len = range[1] + 0, range[2] + 0, range[3] + 0
		table.insert(map.ranges, {s1=src, s2=src + len - 1, d1=dst})
	end
	table.insert(maps, map)
end

for _, map in ipairs(maps) do
	local i = 1
	local seed = seedrngs[i]
	while seed do
		for _, range in ipairs(map.ranges) do
			if not (seed.x2 < range.s1 or seed.x1 > range.s2) then
				if seed.x1 < range.s1 then
					table.insert(seedrngs, {x1=seed.x1,
						x2=range.s1 - 1})
					seed.x1 = range.s1
				end

				if seed.x2 > range.s2 then
					table.insert(seedrngs, {x1=range.s2 + 1,
						x2=seed.x2})
					seed.x2 = range.s2
				end

				seed.x1 = seed.x1 - range.s1 + range.d1
				seed.x2 = seed.x2 - range.s1 + range.d1
				break
			end
		end
		i = i + 1
		seed = seedrngs[i]
	end
end

local min = seedrngs[1].x1
for _, seed in ipairs(seedrngs) do
	if seed.x1 < min then min = seed.x1 end
end
print(min)

File: rust/src/bin/puzzle5-2.rs
fn parse_numbers(s: &str) -> Vec<u64> {
    s.split_ascii_whitespace()
        .map(|s| s.parse::<u64>().unwrap())
        .collect()
}

fn read_line() -> Option<String> {
    Some(std::io::stdin().lines().next()?.unwrap())
}

struct Range {
    dst_start: u64,
    src_start: u64,
    src_end: u64,
}

struct Map {
    ranges: Vec<Range>,
}

fn main() {
    let mut seed_ranges = parse_numbers(read_line().unwrap().split_once(':').unwrap().1)
        .chunks_exact(2)
        .map(|pair| {
            let [src, len] = pair.try_into().unwrap();
            (src, src + len - 1)
        })
        .collect::<Vec<(u64, u64)>>();
    assert_eq!(read_line().unwrap().len(), 0);

    let mut maps = Vec::new();
    while let Some(line) = read_line() {
        assert!(line.ends_with("map:"));
        let mut ranges = Vec::new();
        for line in std::io::stdin().lines().map(|s| s.unwrap()) {
            if line.len() == 0 {
                break;
            }
            let [dst, src, len] = parse_numbers(&line).try_into().unwrap();
            ranges.push(Range {
                dst_start: dst,
                src_start: src,
                src_end: src + len - 1,
            });
        }
        maps.push(Map { ranges });
    }

    for map in &maps {
        let mut i = 0usize;
        while i < seed_ranges.len() {
            let (seed_start, seed_end) = &mut seed_ranges[i];
            let mut new_seed_ranges = Vec::new();
            for range in &map.ranges {
                if *seed_end < range.src_start || *seed_start > range.src_end {
                    continue;
                }
                if *seed_start < range.src_start {
                    new_seed_ranges.push((*seed_start, range.src_start - 1));
                    *seed_start = range.src_start;
                }
                if *seed_end > range.src_end {
                    new_seed_ranges.push((range.src_end + 1, *seed_end));
                    *seed_end = range.src_end;
                }
                *seed_start = *seed_start - range.src_start + range.dst_start;
                *seed_end = *seed_end - range.src_start + range.dst_start;
                break;
            }
            seed_ranges.append(&mut new_seed_ranges);
            i += 1;
        }
    }
    println!("{}", seed_ranges.into_iter().map(|r| r.0).min().unwrap());
}
